gravitee-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/MembershipServiceImpl.java

```diff
@@ -536,47 +540,106 @@ public Metadata findUserMembershipMetadata(List<UserMembership> memberships, Mem
+        );
+    }
+
+    @Override
+    public Page<MemberEntity> getMembersByReferenceAndRole(MembershipReferenceType referenceType, String referenceId, String role, Pageable pageable) {
+        return this.getMembersByReferencesAndRole(referenceType, Collections.singletonList(referenceId), role, pageable);
     }
 
     @Override
     public Set<MemberEntity> getMembersByReferenceAndRole(MembershipReferenceType referenceType, String referenceId, String role) {
-        return this.getMembersByReferencesAndRole(referenceType, Arrays.asList(referenceId), role);
+        return new HashSet<>(
+                this.getMembersByReferencesAndRole(referenceType, Collections.singletonList(referenceId), role, null).getContent()
+        );
     }
 
     @Override
-    public Set<MemberEntity> getMembersByReferencesAndRole(MembershipReferenceType referenceType, List<String> referencesId, String role) {
+    public Set<MemberEntity> getMembersByReferencesAndRole(MembershipReferenceType referenceType, List<String> referenceIds, String role) {
+        return new HashSet<>(
+                this.getMembersByReferencesAndRole(referenceType, referenceIds, role, null).getContent()
+        );
+    }
+
+    @Override
+    public Page<MemberEntity> getMembersByReferencesAndRole(MembershipReferenceType referenceType, List<String> referenceIds, String role, Pageable pageable) {
         try {
-            LOGGER.debug("Get members for {} {}", referenceType, referencesId);
+            LOGGER.debug("Get members for {} {}", referenceType, referenceIds);
             Set<io.gravitee.repository.management.model.Membership> memberships = membershipRepository.findByReferencesAndRoleId(
                     convert(referenceType),
-                    referencesId,
+                    referenceIds,
                     role);
             Map<String, MemberEntity> results = new HashMap<>();
             memberships.stream()
-            .filter(member -> member.getMemberType() == io.gravitee.repository.management.model.MembershipMemberType.USER)
-            .map(this::convertToMemberEntity)
-            .forEach(member -> {
-                String key = member.getId()+member.getReferenceId();
-                MemberEntity existingEntity = results.get(key);
-                if(existingEntity == null) {
-                    results.put(key, member);
-                    existingEntity = member;
-                } else {
-                    Set<RoleEntity> existingRoles = new HashSet<>(existingEntity.getRoles());
-                    existingRoles.addAll(member.getRoles());
-                    existingEntity.setRoles(new ArrayList<>(existingRoles));
-                }
-            });
-            return new HashSet<>(results.values());
+                    .filter(member -> member.getMemberType() == io.gravitee.repository.management.model.MembershipMemberType.USER)
+                    .map(this::convertToMemberEntity)
```

Comment:
I see that we still invoke the `convertToMemberEntity` for each member of the group, one by one. The method `convertToMemberEntity` then fetches the user and fetches each of its roles one by one.
At the end, the sort &  pagination is applied in memory.

I'm afraid it will still continue to be slow with groups having a lot of users no ? 🧐

Reply:
- my first attempts was to paginate directly in repository layer -> unfortunately, we cannot do that because of the work done on roles -> if we query for a page of 10, we can have a result of size 9
- second attempt was to first convertTo member entity, do the work for the roles, and then fetch the user data for the truncated collection (with size of the page in paginate() method) -> the problem is we have to keep the sorting by display name, which is fetched by the user service.


So, what I've done here is to modify the findByIds to have only one call for users information instead of each.
Doing that we have:
- one call to get a list of members
- calls to fetch roles (which will be cached due to your work on /apis)
- one call for the user information of the size of the page (in the case of this page, we will find infos for 10 items)

But I forgot to commit it 😅 , so it is normal it is not clear

Reply:
Ok thx for the information. Here we do not depend on any business logic to calculate something before applying the sort or pagination so I was just wondering if it could be better to manage it on the repository side using an appropriate query.
I know that we support multiple db engines but I can't see any limitation which could prevent us to query and paginate on the db side 🧐 (mongodb => aggregation, jdbc => join).
So, if we want to retrieve all the users belonging to a group, it should be possible to join users with memberships and filter on the targeted group no ?